File: sdks/typescript/src/case.ts
/**
 * Define a full test case in code — everything a `*.skilltest.yaml` file
 * carries — and hand it straight to {@link import("./runner.js").runSkill | runSkill}.
 *
 * ```ts
 * import { runSkill, testCase, boolean, numeric, user } from "@skill-test/sdk";
 *
 * const report = await runSkill(
 *   testCase({
 *     skill: "skills/greeter",
 *     input: "Greet Dr. Smith, who has an appointment today.",
 *     evals: [
 *       boolean("the reply greets Dr. Smith by name"),
 *       numeric("how warm is the tone", { min: 0, max: 10, threshold: 7 }),
 *     ],
 *   }),
 * );
 * if (!report.passed) throw new Error("evals failed");
 * ```
 *
 * A {@link TestCaseInput} is the code twin of the YAML schema (`docs/schema.md`):
 * `skill` resolves relative to the working directory, `input` is the first user
 * message, `user` turns the case multi-turn, `mocks` are the same
 * {@link import("./mock.js").spy | spy} / {@link import("./mock.js").stub | stub}
 * / {@link import("./mock.js").deny | deny} / {@link import("./mock.js").rewrite | rewrite}
 * objects you would pass to `runSkill({ mocks })` (bound for assertions after
 * the run, and referenceable by name from a `called`/`notCalled` eval), and
 * `evals` decide pass/fail. The CLI validates the compiled case, so a malformed
 * one throws a {@link import("./errors.js").SkilltestUsageError | SkilltestUsageError}.
 *
 * Writing cases in code is the recommended approach; YAML files remain fully
 * supported (`runSkill("cases/greet.yaml")`) and are what `discover` collects.
 */
import { SkilltestUsageError } from "./errors.js";
import { type Criterion, ToolMock, type ToolSpy, type WhereCriteria } from "./mock.js";

/** One eval, built with {@link boolean} / {@link numeric} / {@link called} /
 * {@link notCalled}. Opaque: it carries the JSON the CLI validates. */
export interface Eval {
  readonly json: Record<string, unknown>;
}

/** The simulated-user block that makes a case multi-turn (see {@link user}). */
export interface SimulatedUser {
  persona: string;
  doneWhen?: string;
  maxTurns?: number;
}

/** A full test case defined in code — the twin of a `*.skilltest.yaml` file.
 * Pass one (via {@link testCase}) to {@link import("./runner.js").runSkill}. */
export interface TestCaseInput {
  /** Skill directory under test (a dir containing SKILL.md), resolved relative
   * to the working directory. */
  skill: string;
  /** The initial data/prompt handed to the skill as the first user message. */
  input: string;
  /** The evals that decide pass/fail (must be non-empty). */
  evals: Eval[];
  /** Optional report label (defaults to `case`). */
  name?: string;
  /** Present ⇒ multi-turn (see {@link user}). */
  user?: SimulatedUser;
  /** Mock/spy objects for this case — the same builders `runSkill({ mocks })`
   * takes. Bound for assertions after the run; a named one is referenceable
   * from a `called`/`notCalled` eval. */
  mocks?: ToolSpy[];
  /** Force the observation channel even without mocks (implied when `mocks` is
   * non-empty), so the report carries `mock_calls`. */
  spy?: boolean;
}

/** Assert a plain-English `criterion` holds (or, with `expected: false`, that
 * it does not). Scored by the judge against the transcript. */
export function boolean(
  criterion: string,
  options: { expected?: boolean; name?: string } = {},
): Eval {
  const json: Record<string, unknown> = { type: "boolean", criterion };
  json.expected = options.expected ?? true;
  if (options.name !== undefined) json.name = options.name;
  return { json };
}

/** Score `criterion` on the `[min, max]` scale and pass when the score
 * satisfies `comparator` (`>=` `>` `<=` `<`, default `>=`) against `threshold`. */
export function numeric(
  criterion: string,
  options: { min: number; max: number; threshold: number; comparator?: string; name?: string },
): Eval {
  const json: Record<string, unknown> = {
    type: "numeric",
    criterion,
    min: options.min,
    max: options.max,
    threshold: options.threshold,
    comparator: options.comparator ?? ">=",
  };
  if (options.name !== undefined) json.name = options.name;
  return { json };
}

/** Deterministic (no judge): assert the named `mock`/spy observed at least one
 * matching call, or exactly `times`. Reference a spy/stub/… by the `name` you
 * gave it in the case's `mocks`. `where` narrows by input field. */
export function called(
  mock: string,
  options: { times?: number; where?: WhereCriteria; name?: string } = {},
): Eval {
  const json: Record<string, unknown> = { type: "called", mock };
  if (options.times !== undefined) json.times = options.times;
  if (options.where !== undefined) json.where = compileWhere(options.where);
  if (options.name !== undefined) json.name = options.name;
  return { json };
}

/** Deterministic: assert the named `mock`/spy observed **no** matching call
 * (optionally narrowed by `where`). */
export function notCalled(
  mock: string,
  options: { where?: WhereCriteria; name?: string } = {},
): Eval {
  const json: Record<string, unknown> = { type: "not_called", mock };
  if (options.where !== undefined) json.where = compileWhere(options.where);
  if (options.name !== undefined) json.name = options.name;
  return { json };
}

/** A simulated user for a multi-turn case: after each assistant turn the judge
 * plays `persona` until `doneWhen` holds, the skill reports done, or `maxTurns`
 * is reached. */
export function user(
  persona: string,
  options: { doneWhen?: string; maxTurns?: number } = {},
): SimulatedUser {
  return { persona, doneWhen: options.doneWhen, maxTurns: options.maxTurns };
}

/**
 * Build a code-defined case. A convenience identity over {@link TestCaseInput}
 * so `runSkill(testCase({...}))` reads well and infers the eval types; you may
 * also pass a plain object literal to `runSkill` directly.
 */
export function testCase(input: TestCaseInput): TestCaseInput {
  return input;
}

/** @internal Whether a value is a code-defined case (vs a path string). */
export function isTestCaseInput(value: unknown): value is TestCaseInput {
  return typeof value === "object" && value !== null && "skill" in value && "evals" in value;
}

/** @internal The JSON the CLI ingests via `--case-json`: assigns names to the
 * case's mocks (so binding and `called`/`notCalled` references resolve) and
 * turns the spy channel on when any mock/spy is present. */
export function compileCase(input: TestCaseInput): Record<string, unknown> {
  const payload: Record<string, unknown> = {
    skill: input.skill,
    input: input.input,
    evals: input.evals.map((e) => e.json),
  };
  if (input.name !== undefined) payload.name = input.name;
  if (input.user !== undefined) payload.user = compileUser(input.user);
  const decls = compileCaseMocks(input.mocks ?? []);
  if (decls.length > 0) payload.mocks = decls;
  if (input.spy || (input.mocks && input.mocks.length > 0)) payload.spy = true;
  return payload;
}

function compileUser(u: SimulatedUser): Record<string, unknown> {
  const json: Record<string, unknown> = { persona: u.persona };
  if (u.doneWhen !== undefined) json.done_when = u.doneWhen;
  if (u.maxTurns !== undefined) json.max_turns = u.maxTurns;
  return json;
}

function compileWhere(where: WhereCriteria): Record<string, unknown> {
  return Object.fromEntries(
    Object.entries(where).map(([key, value]) => [key, compileWhereCriterion(key, value)]),
  );
}

// A case's `called`/`notCalled` `where` runs hook-side (Rust), so only exact
// strings and the shipped matchers can cross — mirror the mock rule.
function compileWhereCriterion(key: string, criterion: Criterion): unknown {
  if (typeof criterion === "string") return criterion;
  if (
    typeof criterion === "object" &&
    criterion !== null &&
    "kind" in criterion &&
    criterion.kind === "matcher" &&
    criterion.compiled !== undefined
  ) {
    return criterion.compiled;
  }
  throw new SkilltestUsageError(
    `eval \`where\` field \`${key}\` must be an exact string, contains(), or matching() — arbitrary predicates cannot run hook-side`,
  );
}

function compileCaseMocks(mocks: readonly ToolSpy[]): Record<string, unknown>[] {
  const decls: Record<string, unknown>[] = [];
  mocks.forEach((mock, index) => {
    if (mock instanceof ToolMock) {
      decls.push(mock.decl(mock.mockName ?? `__case_mock_${index}`));
    } else if (mock.mockName !== undefined) {
      decls.push(mock.caseDecl(mock.mockName));
    }
  });
  return decls;
}

File: sdks/python/skilltest_sdk/case.py
"""Define a full test case in Python — everything a `*.skilltest.yaml` file
carries — and hand it straight to [`run_skill`][skilltest_sdk.runner.run_skill].

```python
from skilltest_sdk import TestCase, run_skill, boolean, numeric, user

case = TestCase(
    skill="skills/greeter",
    input="Greet Dr. Smith, who has an appointment today.",
    evals=[
        boolean("the reply greets Dr. Smith by name"),
        numeric("how warm is the tone", min=0, max=10, threshold=7),
    ],
)

report = run_skill(case)
assert report.passed
```

A `TestCase` is the code twin of the YAML schema (`docs/schema.md`): `skill`
resolves relative to the working directory, `input` is the first user message,
`user` turns the case multi-turn, `mocks` are the same
[`spy`][skilltest_sdk.mock.spy] / [`stub`][skilltest_sdk.mock.stub] /
[`deny`][skilltest_sdk.mock.deny] / [`rewrite`][skilltest_sdk.mock.rewrite]
objects you would pass to ``run_skill(mocks=...)`` (bound for assertions after
the run, and referenceable by name from a `called`/`not_called` eval), and
`evals` decide pass/fail. The CLI validates the compiled case, so a malformed
one is a loud [`SkilltestUsageError`][skilltest_sdk.errors.SkilltestUsageError],
never a vacuous pass.

Writing cases in code is the recommended approach — it keeps the case, its
mocks, and any deterministic transcript checks in one typed place. YAML files
remain fully supported (`run_skill("cases/greet.yaml")`) and are what the
plugins auto-discover; see the package README.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING, Any

from .mock import Criterion, ToolMock, ToolSpy, _compile_criterion

if TYPE_CHECKING:
    from collections.abc import Sequence

__all__ = [
    "Eval",
    "SimulatedUser",
    "TestCase",
    "boolean",
    "called",
    "not_called",
    "numeric",
    "user",
]


def _drop_none(obj: dict[str, Any]) -> dict[str, Any]:
    """A JSON object with the ``None``-valued keys removed, so optional fields
    are simply absent (the Rust case type applies its own defaults)."""
    return {key: value for key, value in obj.items() if value is not None}


# ---------------------------------------------------------------------------
# Evals
# ---------------------------------------------------------------------------


class Eval:
    """One eval, built with [`boolean`][skilltest_sdk.case.boolean] /
    [`numeric`][skilltest_sdk.case.numeric] / [`called`][skilltest_sdk.case.called]
    / [`not_called`][skilltest_sdk.case.not_called]. Opaque: it carries the
    JSON the CLI validates."""

    def __init__(self, json: dict[str, Any]) -> None:
        self._json = json


def boolean(criterion: str, *, expected: bool = True, name: str | None = None) -> Eval:
    """Assert a plain-English ``criterion`` holds (or, with ``expected=False``,
    that it does not). Scored by the judge against the transcript."""
    return Eval(
        _drop_none({"type": "boolean", "criterion": criterion, "expected": expected, "name": name})
    )


def numeric(
    criterion: str,
    *,
    min: float,
    max: float,
    threshold: float,
    comparator: str = ">=",
    name: str | None = None,
) -> Eval:
    """Score ``criterion`` on the ``[min, max]`` scale and pass when the score
    satisfies ``comparator`` (``>=`` ``>`` ``<=`` ``<``) against ``threshold``."""
    return Eval(
        _drop_none(
            {
                "type": "numeric",
                "criterion": criterion,
                "min": min,
                "max": max,
                "threshold": threshold,
                "comparator": comparator,
                "name": name,
            }
        )
    )


def called(
    mock: str,
    *,
    times: int | None = None,
    where: dict[str, Criterion] | None = None,
    name: str | None = None,
) -> Eval:
    """Deterministic (no judge): assert the named ``mock``/spy observed at least
    one matching call, or exactly ``times``. Reference a
    [`spy`][skilltest_sdk.mock.spy]/[`stub`][skilltest_sdk.mock.stub]/… by the
    ``name=`` you gave it in the case's ``mocks``. ``where`` narrows by input
    field (exact string or [`contains`][skilltest_sdk.mock.contains]/
    [`matching`][skilltest_sdk.mock.matching])."""
    json: dict[str, Any] = {"type": "called", "mock": mock}
    if times is not None:
        json["times"] = times
    if where:
        json["where"] = {key: _compile_criterion(key, value) for key, value in where.items()}
    if name is not None:
        json["name"] = name
    return Eval(json)


def not_called(
    mock: str,
    *,
    where: dict[str, Criterion] | None = None,
    name: str | None = None,
) -> Eval:
    """Deterministic: assert the named ``mock``/spy observed **no** matching
    call (optionally narrowed by ``where``)."""
    json: dict[str, Any] = {"type": "not_called", "mock": mock}
    if where:
        json["where"] = {key: _compile_criterion(key, value) for key, value in where.items()}
    if name is not None:
        json["name"] = name
    return Eval(json)


# ---------------------------------------------------------------------------
# Simulated user (multi-turn)
# ---------------------------------------------------------------------------


@dataclass
class SimulatedUser:
    """The simulated-user block that makes a case multi-turn. Build with
    [`user`][skilltest_sdk.case.user]."""

    persona: str
    done_when: str | None = None
    max_turns: int | None = None

    @property
    def _json(self) -> dict[str, Any]:
        return _drop_none(
            {
                "persona": self.persona,
                "done_when": self.done_when,
                "max_turns": self.max_turns,
            }
        )


def user(
    persona: str, *, done_when: str | None = None, max_turns: int | None = None
) -> SimulatedUser:
    """A simulated user for a multi-turn case: after each assistant turn the
    judge plays ``persona`` until ``done_when`` holds, the skill reports done,
    or ``max_turns`` is reached."""
    return SimulatedUser(persona=persona, done_when=done_when, max_turns=max_turns)


# ---------------------------------------------------------------------------
# The case
# ---------------------------------------------------------------------------


@dataclass
class TestCase:
    """A full test case defined in code — the twin of a `*.skilltest.yaml` file.

    Pass one to [`run_skill`][skilltest_sdk.runner.run_skill] /
    [`stream_skill`][skilltest_sdk.stream.stream_skill] in place of a path.
    """

    # Its name starts with `Test`, but this is not a pytest test class — tell
    # pytest not to try to collect it (avoids a PytestCollectionWarning wherever
    # a user imports it into a test module). Unannotated, so `@dataclass` ignores
    # it as a field.
    __test__ = False

    #: Skill directory under test (a dir containing SKILL.md), resolved relative
    #: to the working directory.
    skill: str | Path
    #: The initial data/prompt handed to the skill as the first user message.
    input: str
    #: The evals that decide pass/fail (must be non-empty).
    evals: Sequence[Eval]
    #: Optional report label (defaults to ``case`` when omitted).
    name: str | None = None
    #: Present => multi-turn (see [`user`][skilltest_sdk.case.user]).
    user: SimulatedUser | None = None
    #: Mock/spy objects for this case — the same builders ``run_skill(mocks=)``
    #: takes. Bound for assertions after the run; a named one is referenceable
    #: from a `called`/`not_called` eval.
    mocks: Sequence[ToolSpy] = field(default_factory=tuple)
    #: Force the observation channel even without mocks (implied when ``mocks``
    #: is non-empty), so the report carries ``mock_calls``.
    spy: bool = False

    def _compile(self) -> dict[str, Any]:
        """The JSON the CLI ingests via ``--case-json``. Assigns names to the
        case's mocks (so binding and `called`/`not_called` references resolve)
        and turns the spy channel on when any mock/spy is present."""
        payload: dict[str, Any] = {
            "skill": str(self.skill),
            "input": self.input,
            "evals": [e._json for e in self.evals],
        }
        if self.name is not None:
            payload["name"] = self.name
        if self.user is not None:
            payload["user"] = self.user._json
        decls = _compile_case_mocks(self.mocks)
        if decls:
            payload["mocks"] = decls
        if self.spy or self.mocks:
            payload["spy"] = True
        return payload


def _compile_case_mocks(mocks: Sequence[ToolSpy]) -> list[dict[str, Any]]:
    """Compile a case's ``mocks`` into declarations: every intercepting mock
    (stub/deny/rewrite) becomes an action declaration; a **named** spy becomes a
    no-action declaration so a `called`/`not_called` eval can reference it.
    Unnamed spies contribute nothing here — they filter locally after the run —
    but still turn the channel on via the case's ``spy`` flag."""
    decls: list[dict[str, Any]] = []
    for index, mock in enumerate(mocks):
        if isinstance(mock, ToolMock):
            decls.append(mock._decl(mock._user_name or f"__case_mock_{index}"))
        elif mock._user_name is not None:
            decls.append(mock._case_decl(mock._user_name))
    return decls
